File: go/datastore/block_writer.go
package datastore

import (
	"bytes"
	"encoding/binary"
)

type BlockWriter struct {
	opts     *TabletOptions
	buf      *bytes.Buffer
	count    uint
	restarts []uint32
	firstKey []byte
	prevKey  []byte
}

func NewBlockWriter(opts *TabletOptions) *BlockWriter {
	// initialize buf to 2*BlockSize to minimize resizes
	buf := bytes.NewBuffer(make([]byte, 0, 2*opts.BlockSize))
	return &BlockWriter{opts: opts, buf: buf}
}

// number of bytes of prefix in common between two byte slices
func commonPrefix(bin1 []byte, bin2 []byte) uint {
	length := len(bin1)
	if len(bin2) < length {
		length = len(bin2)
	}

	var count uint
	for i := 0; i < length; i++ {
		if bin1[i] == bin2[i] {
			count++
		}
	}

	return count
}

func (b *BlockWriter) Append(key []byte, value []byte) {
	if b.buf.Len() == 0 {
		b.firstKey = key
	}

	var shared uint
	if b.count%b.opts.KeyRestartInterval == 0 {
		b.restarts = append(b.restarts, uint32(b.buf.Len()))
	} else {
		shared = commonPrefix(b.prevKey, key)
	}

	writeUint(b.buf, shared)
	writeKv(b.buf, key[shared:len(key)], value)
	b.prevKey = key
	b.count += 1
}

func (b *BlockWriter) Size() uint32 {
	return uint32(b.buf.Len() + 4*len(b.restarts) + 4)
}

func (b *BlockWriter) Finish() (firstKey []byte, buf []byte) {
	for _, restart := range b.restarts {
		binary.Write(b.buf, binary.BigEndian, restart)
	}

	binary.Write(b.buf, binary.BigEndian, uint32(len(b.restarts)))

	return b.firstKey, b.buf.Bytes()
}

func (b *BlockWriter) Reset() {
	b.firstKey = nil
	b.buf.Reset()
}

File: java/src/main/java/com/thefactory/datastore/BlockWriter.java
package com.thefactory.datastore;

import org.msgpack.MessagePack;
import org.msgpack.packer.Packer;

import java.io.*;
import java.nio.ByteBuffer;
import java.util.Arrays;

public class BlockWriter {
    private TabletOptions opts;

    private Packer packer;

    private ByteArrayOutputStream buf;
    private ByteArrayOutputStream restarts;

    private byte[] prevKey;
    private byte[] firstKey;
    private int keyCount;

    public BlockWriter(TabletOptions opts) {
        this.opts = opts;
        this.buf = new ByteArrayOutputStream(2*opts.blockSize);

        MessagePack msgpack = new MessagePack();
        this.packer = msgpack.createPacker(this.buf);

        this.restarts = new ByteArrayOutputStream();
        this.keyCount = 0;
    }

	public void append(byte[] key, byte[] value) throws IOException {
        if (buf.size() == 0) {
            firstKey = Arrays.copyOf(key, key.length);
        }

        int shared = 0;

        if (keyCount % opts.keyRestartInterval == 0) {
            writeInt(restarts, buf.size());
        } else {
            shared = Utils.commonPrefix(prevKey, key);
        }

        packer.write(shared);
        packer.write(key, shared, key.length-shared);
        packer.write(value);

        this.prevKey = key;
        this.keyCount += 1;
    }

    private void writeInt(OutputStream out, int pos) throws IOException {
        out.write((byte) (pos >> 24));
        out.write((byte) (pos >> 16));
        out.write((byte) (pos >> 8));
        out.write((byte) (pos));
    }

    public int size() {
        // include 4 bytes for the restarts count
        return buf.size() + restarts.size() + 4;
    }

    public byte[] getFirstKey() {
        return firstKey;
    }

    public byte[] finish() throws IOException {
        buf.write(restarts.toByteArray());
        writeInt(buf, restarts.size() / 4);
        return buf.toByteArray();
    }

    public void reset() {
        buf.reset();
        restarts.reset();
        prevKey = null;
        firstKey = null;
        keyCount = 0;
    }
}
